src/odometry_publisher.cpp

```diff
@@ -0,0 +1,215 @@
+    leftPos = newLeftPos;
+    rightPos = newRightPos;
+    double dLeftDistance = dLeftPos * base->getWheelDiameter() / 2.0d;
+    double dRightDistance = dRightPos * base->getWheelDiameter() / 2.0d;
+    std::cout << "dRightDistance: " << dRightDistance << std::endl;
+    leftDistance += dLeftDistance;
+    rightDistance += dRightDistance;
+
+    // Calculate radius
+    double radius;
+    if(myAbs(dLeftDistance - dRightDistance) < 0.00001d) {
+        radius = 0;
+    } else {
+        radius = base->getWheelBase() / 2.0 * ((dLeftDistance + dRightDistance) / (dLeftDistance - dRightDistance));
+        if (radius == 0){
+        	turnInPlace = true;
+        }
+    }
+    if(radius < 0) radius = -1*radius;
+
+    // Calculate theta from the radius
+    if(dLeftDistance > 0 || dLeftDistance < 0) {
+        if((dRightDistance < dLeftDistance && dLeftDistance > 0) || (dLeftDistance < 0 && dRightDistance > dLeftDistance)) {
+            dTheta = dLeftDistance / (radius + (base->getWheelBase() /2.0));
+            if (logging) std::cout << "BUITENBOCHT" << std::endl;
+        } else {
+            dTheta = dLeftDistance / (radius - (base->getWheelBase() /2.0));
+            if (logging) std::cout << "BINNENBOCHT" << std::endl;
+        }
+    } else {
+        dTheta = dRightDistance / (radius + (base->getWheelBase() /2.0));
+    }
+
+    if (dLeftDistance < 0 || dRightDistance < 0) {
+        dTheta = -1*dTheta;
+    }
+    if(radius == 0 && !turnInPlace) {
+       dTheta = 0;
+    }
+
+	//if (dLeftDistance > 0 && dRightDistance < 0){
+    if(dLeftDistance > 0 && dRightDistance < 0 && !turnInPlace) {
+       dTheta = -1*dTheta;
+    }
+
+    // Calculate the difference in dx and dy
+    dx = radius*cos(dTheta) - radius;
+    dy = radius*sin(dTheta);
+
+    // Corner to the right, invert angle and dx
+    //if(dRightDistance < dLeftDistance) {
+    if(dRightDistance < dLeftDistance && !turnInPlace) {
+        dTheta = -1*dTheta;
+        dx = -1*dx;
+    }
+
+    // One of the sides driving backwards, invert both axis
+    if (dLeftDistance < 0 || dRightDistance < 0) {
+        dx = -1*dx;
+        dy = -1*dy;
+    }
+
+    // Calculate odom coordinates
+    x += dx*cos(th) - dy*sin(th);
+    y += dx*sin(th) + dy*cos(th);
+    th += dTheta;
+
+    // Log all the things
+    if(logging) {
+        std::cout << "Leftpos: " << leftPos << std::endl;
+        std::cout << "Rightpos: " << rightPos << std::endl;
+        std::cout << "Left distance: " << leftDistance << std::endl;
+        std::cout << "Right distance: " << rightDistance << std::endl;
+        std::cout << "Radius: " << radius << std::endl;
+        std::cout << "dTheta: " << dTheta << std::endl;
+        std::cout << "dx: " << dx << std::endl;
+        std::cout << "x " << x << std::endl;
+        std::cout << "y " << y << std::endl;
+        std::cout << "th " << th << std::endl;
+        std::cout << "th " << 360 * (th / (2 * 3.1415)) << std::endl;
+        std::cout << "\n";
```

Comment:
Cool logging, but may I suggest
(A) replace every std::endl with '\n'
(B) replace the single final '\n' with cout << '\n' << endl.
TLDR, \n is fast, endl flushes the output buffer. So shit lots of newlines, then when done flush the toilet.